### ISSUE TITLE
Any way to add the ability to find what *combination* of existing tokens best matches a token?

### ISSUE THREAD

**Issue by enn-nafnlaus** [tkalayci71/embedding-inspector#14]
It's occurred to me that this could be an incredibly valuable tool, not just for usage, but for also making a key point.

Right now, this (awesome) plugin will tell you what tokens are "most similar" to a given token, but not how to mix tokens to best recreate a given token. This feature would, of course, be useful on its own, as people could change the balance to emphasize or deemphasize various factors.

However, it would in practice be much, much more. 

Anti-AI-artists are obsessed with the notion of putting the genie back in the bottle.  And their primary goal is that, "_If only we can get a bunch of artists excluded from the dataset, then nobody will be able to produce the styles they want and this whole thing will all go away!_"  Except, of course, that's not the case. The checkpoints don't learn each artist's style from scratch, only the minor statistical differences between them and the rest of the collective dataset; 99% of them could easily be excluded but artwork similar to them would still exist somewhere in the latent space.

If one could use this extension to find combinations of tokens that best recreate a token in latent space, one could take an embedding trained on a given artist's style, with a checkpoint where that artist was **not** part of the training data, and find a handful of existing tokens in that checkpoint which, when mixed, recreate that artist's style.  Poof, the artist can be found in the checkpoint, even though they weren't in its training dataset.

One could of course object to the fact that a trained embedding was used to find them, but that objection goes nowhere.  A person could readily have found their style in the latent space manually - it just would have taken longer.  Such a tool would simply speed up the process.  And indeed, once a weighted combination of tokens that matches an artist's style is found, I suspect that it would be relatively stable between checkpoints.

I know that finding combinations of tokens isn't as computationally simple as finding individual tokens - but you don't have to exhaust the whole search space looking for a "perfect" solution, just a "good" solution. A lazy solution would just be to sort tokens by similarity, then for X cycles (or until interrupted) random-select Y tokens, selection being weighted by similarity to the target token, and run scipy.optimize.minimize on them to find the optimal weightings.  You could then keep and present the ~10 best combinations of tokens.

Thinking about it, I imagine that a faster-converging algorithm (though still far from optimal) might be, say, for each cycle to weighted-random-select Y * 1000 tokens, run scipy.optimize.minimize to find the weights, throw away the bottom Y*900 tokens, re-run scipy.optimize.minimize, throw away the bottom Y * 90 tokens, re-run scipy.optimize.minimize, throw away the bottom Y * 9 tokens,then re-run scipy.optimize.minimize one last time before comparing it to other previously found token combinations/weights.

**Comment by tkalayci71** [comment]
Hi, thanks for you message, I'm not really interested in that artist debate, but what you're describing was kind of the idea behind this extension; edit a TI embedding as a mixture of internal embeddings.  However it didn't seem to work, maybe I didn't do it right (was first time I wrote the optimization loop, I'm new with all this), didn't work unless I mixed like 1000 embeddings, and that's no better than editing with 768 sliders, so I had abandoned that idea. It should be simple enough for someone more experienced than me, or I may retry with your convergence idea sometime. Thanks.

**Comment by enn-nafnlaus** [comment]
I don't have experience with the internals behind tokens, but I do have extensive experience with scipy.optimize.minimize, so if you'd like me to code up an algorithm for you which you could integrate, I'd be glad to.  I'd just need to know what sort of data structure I should use for the tokens (and for testing it, roughly how many simulated tokens I should create... or contrarily, I could use real data for testing if I could get it in some easy-to-load format).

**Comment by tkalayci71** [comment]
That would be great. I can create a quick project, probably within 24-hours, and post here.

**Comment by tkalayci71** [comment]
This is my first collaboration, I hope this works..

https://github.com/tkalayci71/recreate_emb

So, your algorithm would be in Calc_best function # ... optimization loop here ...
which should calculate best_indexes and best_weights

best_indexes are indexes to embeddings in all_embs which has shape of torch.Size([49408, 768])
target_emb is the target embedding which has shape of torch.Size([768])

I included a loss function, not entirely sure it's correct.

Feel free to optimize/change any other part, I'm new to torch and python as well.
Let me know if clarification is needed, thanks.

**Comment by enn-nafnlaus** [comment]
Excellent.  :)   I hadn't checked in on this thread all day, but just instead went off and wrote an optimization function  ;)  I'm now testing it on a bunch of different parameters to see what finds matches quickest (at least with my purely random test data).  Once I've gotten the answer to that I'll adapt it to the above  :)

**Comment by enn-nafnlaus** [comment]
So, the tests are still running, but I looked over your code, and it should be easy to adapt my code in, so long as PyTorch Tensors respond to mathematical operations the same as numpy Arrays.  :)

As an example of the sort of similarities I'm seeing in my testing, BTW, I'm getting results like (with 100k random-generated tokens of width=25), with a couple minutes of calculation time:

```
Score: 0.31363324248531266
this_emb:   [ 0.5203 -0.8746 -0.4381 -0.43    0.9315 -0.5075 -0.6138 -0.4101 -0.3603  0.3431 -0.1034 -0.6183 -0.0013  0.4069  0.4594  0.5209 -0.7046  0.1132  0.7561 -0.8978 -0.809   0.9901  0.5849 -0.0907  0.1304]
target_emb: [ 0.7285 -0.7652 -0.5333 -0.4994  0.9587 -0.3621 -0.5526 -0.5002 -0.5537  0.3203 -0.3486 -0.7426 -0.0953  0.5919  0.2899  0.4182 -0.657   0.174   0.7458 -0.8282 -0.7604  0.9873  0.5748 -0.0664  0.1607]

  -0.07655003387267577: [ 0.4077  0.8366  0.7832  0.3551 -0.4955  0.8245 -0.0303  0.916  -0.1101 -0.177  -0.1886  0.5776  0.901   0.1656  0.9187 -0.9719 -0.1263 -0.532  -0.2469  0.2097  0.6245  0.1089 -0.4226 -0.514   0.3407]
  0.09113732917848788: [-0.3507  0.694   0.9366  0.264   0.5035  0.5431  0.7815 -0.3169  0.3152 -0.0065  0.4498 -0.8925  0.7621  0.0214  0.4867 -0.2271  0.9563 -0.2423  0.647   0.8634  0.9312 -0.0659 -0.2007  0.0456 -0.3408]
  0.1123886612854348: [-0.0185 -0.4557 -0.3672 -0.6957 -0.7915  0.1152  0.8039  0.4613  0.5809 -0.6668  0.1302 -0.0745  0.3475  0.3892 -0.0284 -0.3809 -0.676   0.8847  0.3965 -0.5968  0.3857  0.0062 -0.5806  0.6537 -0.4045]
  -0.1605896245720823: [ 0.6781  0.8989  0.7505 -0.4415 -0.6897  0.0911 -0.8998  0.9912  0.3324  0.4956  0.5879 -0.2121  0.0527 -0.0162  0.2446 -0.6984  0.9477 -0.1286  0.5963  0.6527  0.2961 -0.9214  0.4098  0.7831  0.4562]
  -0.1703053241529672: [ 0.2387  0.6345 -0.6584  0.5285 -0.7138 -0.3179  0.2821 -0.7459  0.0273  0.0997  0.7302 -0.1698  0.1612 -0.6891 -0.2677  0.0216  0.461   0.5514 -0.7626  0.3391  0.871   0.8377 -0.293  -0.89   -0.2396]
  -0.269643966010229: [-0.0747  0.6998 -0.1214 -0.2582  0.0748  0.7236 -0.0913  0.4083  0.6362  0.1491 -0.0899  0.2194 -0.2428 -0.4214  0.808  -0.8203  0.6624  0.2001 -0.5792  0.1845 -0.3924 -0.9067 -0.954  -0.0475  0.9783]
  -0.3056135293362874: [-0.856   0.0616  0.1784  0.8514 -0.7031  0.5032  0.9318 -0.0299  0.1131 -0.1553 -0.4256  0.2718 -0.1907 -0.6006 -0.6958 -0.3332  0.1559 -0.7162 -0.3064  0.6925 -0.1125 -0.8698  0.1444 -0.0765  0.4491]
  -0.3679436150034775: [-0.8507 -0.8622 -0.1762  0.7937 -0.0904 -0.1144  0.4954 -0.1245  0.0525  0.2443 -0.1373  0.9037  0.425   0.4322 -0.5101 -0.8513  0.653   0.7499 -0.4132  0.6662  0.8265  0.1034 -0.6988  0.0207 -0.2251]
  -0.37191568145099896: [-0.7952  0.6161  0.6611  0.0156 -0.7524 -0.273   0.219   0.8945  0.5519 -0.5856 -0.3381 -0.7943  0.1354  0.5873 -0.3064 -0.2241 -0.1619 -0.8601  0.23    0.5854  0.9622 -0.6606 -0.3471  0.6904 -0.5556]
  -0.4839902265889707: [ 0.3211  0.9314  0.436  -0.3282 -0.4059  0.7048  0.7238 -0.1179 -0.0538 -0.7573  0.5895  0.7605 -0.1612 -0.6905 -0.3839  0.6572  0.1827  0.2903 -0.5853  0.0135  0.3548 -0.5795 -0.1087  0.0108 -0.779 ]
```

**Comment by enn-nafnlaus** [comment]
Question: how do you want to manage how much time to give the algo to find good matches?  First I just had it in a _while True_ loop.  Now I have it set up to run for a fixed number of seconds.  But really I could do it any way that would be desirable.  Should I have some sort of check for some global "interrupt" variable for the user to be able to interrupt it, but otherwise run for a fixed length of time (or indefinitely)?

**Comment by tkalayci71** [comment]
Limiting either number of steps or seconds would work (and if loss is very good, it could exit early), which would be selected by user. UI may remain frozen until it finishes so I think speed is important. But most importantly, recreated&saved embedding should produce similar images as target embedding, as I mentioned I couldn't achieve this by mixing 10 embeddings.

**Comment by enn-nafnlaus** [comment]
Still working on the conversion from numpy to pytorch.  Torch scalars appear to be like numpy arrays that complain a lot  :Þ

**Comment by tkalayci71** [comment]
No rush, take your time. I can check tomorrow.

**Comment by enn-nafnlaus** [comment]
If you want to save me a little time, if you can tell me why this:

  _new_token_group = torch.Tensor([pick_token() for i in range(cur_tokens_per_group)])_

... where the inner list looks like:

_[tensor([-0.5073,  0.8606,  0.7855, -0.8169, -0.1019,  0.6303, -0.9234,  0.9610,  -0.6293, -0.4201, -0.2713,  0.1852,  0.1191, -0.1080,  0.4534, -0.4329, -0.3420,  0.8214,  0.8363, -0.1138,  0.9477, -0.2424, -0.3107, -0.7516, -0.2365]),
tensor([ 0.4936, -0.9110, -0.0492, -0.8396,  0.9987, -0.9945,  0.4225, -0.9166, 0.9413,  0.6945, -0.5910, -0.9578,  0.3544,  0.6114,  0.3778, -0.0870, 0.6657,  0.6075, -0.0216, -0.0328,  0.2348,  0.6721,  0.7451,  0.0892, -0.8118]), .... ]_

would complain about:

_ValueError: only one element tensors can be converted to Python scalars_

... then by all means let me know!  :)

**Comment by enn-nafnlaus** [comment]
So, I got it to work with Torch... kinda.  I tested through getting the first sets of results out of scipy.optimize.minimize.  But the Torch version runs over five times slower than the numpy version  :Þ  

I'm attaching the code here for you to see:

[minimize.py.txt](https://github.com/tkalayci71/embedding-inspector/files/10329745/minimize.py.txt)

If you can see any obvious ways to improve the performance of the torch version, by all means, let me know.  Otherwise I think we should convert the torch Tensors into numpy arrays and just do the optimization in numpy. 

I've modified the code to simultaneously accumulate solutions for all MAX_SIMILAR_EMBS and below, rather than just MAX_SIMILAR_EMBS (e.g. the best two embeddings, the best three, the best four, etc... all simutaneously), because it doesn't affect performance that much when you're starting off by narrowing down like a thousand embeddings.  I've thusfar only tested with my very-narrow 25-wide arrays; right now I'm running tests on 768-wide arrays, with a total count of 49408, to match the actual data. I'll probably need to do a lot of tests overnight to find the optimal parameters. You can see where the parameters are set on lines 20-25.

**Comment by tkalayci71** [comment]
Well, issues:
1) I only got errors, it's expecting command line arguments, what string should I use?
2) You haven't yet used real data I provided. I would very much like to see proof of concept before going further, I mean image results comparing original vs recreated embedding.
3) I'd prefer if we used torch only (I assume it will be faster and optionally be able to use GPU). But, we can convert to torch later, for now it's fine to convert inputs to numpy at start and back to torch at end.
4) I haven't had time to analyze your code, I'd have to understand every line of it, but I'd rather do that after you finalize your algorithm.
5) One thing I noticed is you don't use CosineSimilarity. But with real data, absolute values are not so important, magnitude can be scaled, and a constant can be added, the embedding still works. Still, the algorithm is up to you, as long as it works.

**Comment by enn-nafnlaus** [comment]
> Well, issues:
> 
> 1. I only got errors, it's expecting command line arguments, what string should I use?

As I mentioned, lines 20-25 in the code describe what they're being used for and give examples of possible defaults. An example might be something like "_./minimize.py 1	10	200,50,20,7,3,2,1.5,1.2,1	4	100	0"

I have it take all those parameters as command-line arguments for now to make it easy to benchmark the performance with different parameter settings.

> 2. You haven't yet used real data I provided.

That's because it's Tensors and I only got (disappointing) initial Tensor support in right before I went to bed (and also didn't yet have a plan as to whether to use it as Tensors or to convert it into numpy arrays first).  I plan to try to incorporate it this evening.

> 3. I'd prefer if we used torch only (I assume it will be faster and optionally be able to use GPU). But, we can convert to torch later, for now it's fine to convert inputs to numpy at start and back to torch at end.

I think it's best for me to try to maintain support for both np arrays and torch Tensors, and for now to use np arrays.  So I'll do that.  It should also be possible to add in threading for np arrays to make them even faster.

> 4. I haven't had time to analyze your code, I'd have to understand every line of it, but I'd rather do that after you finalize your algorithm.

Naturally  :)

> 5. One thing I noticed is you don't use CosineSimilarity. But with real data, absolute values are not so important, magnitude can be scaled, and a constant can be added, the embedding still works. Still, the algorithm is up to you, as long as it works.

Funny thing, as I was going to sleep last night that occured to me that I should be doing that, but then I promptly forgot it, so thanks for reminding me.  ;)  I'll implement that this evening too.

**Comment by tkalayci71** [comment]
All is well then. Keep me posted.

**Comment by enn-nafnlaus** [comment]
Update:

[minimize.py.txt](https://github.com/tkalayci71/embedding-inspector/files/10333830/minimize.py.txt)

* Now minimizes CosineDistance
* Now loads the sample embeddings and runs on those.
* Speed fix for Tensors. Haven't verified that they're faster than numpy yet, but it does seem to have improved performance significantly.
* Partial threading implementation for Numpy in place. Incomplete.  I stopped work on it when I found said speed fix for Tensors.
* Algorithmic change.  Haven't verified that it's better yet.
* Unified everything into a class so that you just have to call a couple class functions to get a return value. Sample usage at the bottom.  Sample commandline arguments: ./minimize.py 1   10      17,7,3,2,1.5,1.2,1              8       100     0       6

TODO:
 * Lots of testing and optimization to find the fastest algorithm and settings for convergence. I'm not at all happy with the present convergence performance (to the point that I actually suspect a bug).
 * At some point, if working with tensors, we'd want to run these on the GPU.

At this point, I think the main architecture is down, and now it's primarily just tuning and bughunting.

**Comment by tkalayci71** [comment]
Nice progress, thanks. However, again issues:

1) It's not working. I loaded buddha-statue.bin and run your code, entered the results manually into mixer and obtained the recreated embedding. The images it generates has no resemblence to buddha statues, is just noise. Also, when I inspect it, the vector does not match your results, see below:

this printed best_emb_groups[9]:

[(0.8125810027122498,
  tensor([-1.2562e-01, -2.3056e-02,  5.4965e-02,  2.8526e-02, -8.2620e-02,
         2.2338e-02, -1.3886e-02,  2.9120e-01,  4.8686e-02, -1.5845e-01,
.......
 [(-1.0891472211662097, 26722),
   (-1.303380587118969, 24892),
   (1.5492051717174609, 6564),
   (1.5536558035262074, 9233),
   (1.6113597575386058, 14767),
   (1.806655090326269, 1350),
   (2.6286666990188228, 18402),
   (3.0007243070964864, 1964),
   (3.154699911268092, 861),
   (3.47630798986664, 60)])]

this is my mix:
+ inappropriate</w>(26722) x -1.0891472211662097
+ kham(24892) x -1.303380587118969
+ characters</w>(6564) x 1.5492051717174609
+ raft</w>(9233) x 1.5536558035262074
+ trace</w>(14767) x 1.6113597575386058
+ ny</w>(1350) x 1.806655090326269
+ duke(18402) x 2.6286666990188228
+ follow(1964) x 3.0007243070964864
+ go</w>(861) x 3.154699911268092
+ ](60) x 3.47630798986664
Saved "embeddings/bbb.bin"
Reloading all embeddings

and its vector:
Vector[0] = tensor([-0.0024,  0.0320, -0.0700,  ..., -0.2076, -0.1916,  0.1468])

-----------------

2) I'd like to remind a few things: vector size is 768 for SD1 and 1024 for SD2 , we need to be flexible there (I haven't yet uploaded embeddings_1024.pt file). Also, each embedding can have multiple vectors (like 8, 16 or more). The waiting time will be multiplied by the vector count, so I'm concerned about speed. I don't know how to do it non-blocking way, without freezing webui, are you by any chance familiar with the gradio framework? Or maybe this can be a standalone app, we can hijack/activate venv from webui. What do you think?

**Comment by enn-nafnlaus** [comment]
As expected. 

1) As I wrote, I need to work a lot on optimizing convergence - indeed, the last change seemed to make it worse, and I suspect there may be outright bugs in the code with that regard.

2) As I wrote, the IDs are getting shuffled up, by both the sorting process and by the elimination of empty embeddings.  I need to add a mapping back to the old IDs.

So there's nothing at all surprising about your results.  :)

Thanks for the note re: #2.  Vector size is an easy fix. What's the format of multi-vector embeddings? How are they used?  Like, how would I compare a multi-vector embedding to a single-vector one? How do I weigh and add them, or really, perform any mathematical or logical operations on them when the sizes aren't the same?

**Comment by tkalayci71** [comment]
OK, I can't seem to find where you wrote IDs are getting shuffled, and I don't know what you mean by empty embeddings, that's all your part... But please load buddha-statue.bin and recreate the embedding and test the image it generates, that's the part I care about.

2) Multi-vector embeddings just contain more than one vector, format is when you load there's an extra dimension (I think). I don't see a way other than processing each vector separately.

**Comment by enn-nafnlaus** [comment]
Hmm, I must have dreamed that I mentioned the shuffling/ID issue.  Basically, the algorithm is sorting the embeddings.  So what was the embedding at index 0 is no longer index 0, it's somewhere else.  Same with index 1, and 2, and 3... etc.

Empty embeddings: all values of the tensor are zero.  I don't know why they exist, but they do, there's something like 130 of them, and they cause errors if I don't remove them (as well as wasting time if they get selected).

**Comment by tkalayci71** [comment]
I have another question: when your algorithm picks 10 best embs to mix, are they different ones than most similar 10? I mean otherwise why bother checking all. Also, I think outputting just 10 best embs (or whatever reasonable number user chooses) will be enough, isn't it wasteful to calculate best 9, 8, 7.., etc?

**Comment by tkalayci71** [comment]
I have updated 

https://github.com/tkalayci71/recreate_emb

includes embeddings_1024.bin, example embedding files with different vector size and counts, and functions in main.py that can load/save them. All you need to do your magic in Calc_best function. Let me know when you have image results, thanks.

**Comment by enn-nafnlaus** [comment]
ED: Whoope.  Ran overnight, somehow optimized to a score of "-9.090311447417174e+17".  Aka, while working to minimize scores, it got to a hugely **negative** score.  Better figure out how that happened, I've not seen that before!  Probably due to one of the penalty systems.

**Comment by tkalayci71** [comment]
OK I'm gonna close this issue for now I've lost interest and gonna take a break.
I've also made the other repo private, I hope you downloaded the files.
Also, if you wish, take a look at embedding-mixer, could be useful if your code can output a compatible text format.
Take care!

**Comment by enn-nafnlaus** [comment]
Wow.  Sorry, I thought you wanted to be kept informed of progress? I guess I shouldn't have?  And sorry it took more than 24 hours to complete a project that's nearly as many lines of code as your entire extension, and more complex of a task...  it's not very motivating to finish when the task you've worked for several days just gets closed on you  :Þ

**Comment by tkalayci71** [comment]
Definitely there was a miscommunication between us and it went sour, sorry. But you don't need me to complete this project and it was your vision in the first place. Good luck.

**Comment by tkalayci71** [comment]
using tokenizer:

```
from transformers import CLIPTokenizer
tokenizer = CLIPTokenizer.from_pretrained('openai/clip-vit-large-patch14')
test_id = 2368 # 22979
name_utf8 = tokenizer.decoder.get(test_id)
byte_array_utf8 = bytearray([tokenizer.byte_decoder[c] for c in name_utf8])
name = byte_array_utf8.decode("utf-8", errors='backslashreplace')
print(name)
```

or,  download  https://huggingface.co/CompVis/stable-diffusion-v1-4/blob/main/tokenizer/vocab.json

```
import json
f= open('vocab.json', encoding="utf-8")
encoder = json.load(f)
f.close()
decoder = {v: k for k, v in encoder.items()}
test_id = 2368 #22979
name_utf8 = decoder[test_id]
byte_array_utf8 = bytearray([tokenizer.byte_decoder[c] for c in name_utf8])
name = byte_array_utf8.decode("utf-8", errors='backslashreplace')
print(name)

```

**Comment by tkalayci71** [comment]
```
from transformers import CLIPTokenizer
tokenizer = CLIPTokenizer.from_pretrained('openai/clip-vit-large-patch14')
test_id = 2368 # 22979
name_utf8 = tokenizer.decoder.get(test_id)
byte_array_utf8 = bytearray([tokenizer.byte_decoder[c] for c in name_utf8])
name = byte_array_utf8.decode("utf-8", errors='backslashreplace')
print(name)
```

or, download https://huggingface.co/CompVis/stable-diffusion-v1-4/blob/main/tokenizer/vocab.json

```

import json
f= open('vocab.json', encoding="utf-8")
encoder = json.load(f)
f.close()
decoder = {v: k for k, v in encoder.items()}
test_id = 2368 #22979
name_utf8 = decoder[test_id]
print(name_utf8)
```
this way needs utf8 decoding which I'm not sure how to do

**Comment by enn-nafnlaus** [comment]
Thanks for that. I'm finally to the point of having a result I'd like to try (though there's still a lot more to optimize in terms of performance and convergence):

  -2.172365665435791 * 33961 [49186] "cree"
  -1.893974781036377 * 25892 [49191] "pasadena</w>"
  -1.8557159900665283 * 1061 [49164] "ish</w>"
  -1.7042527198791504 * 25705 [48530] "informal</w>"
  -1.6060261726379395 * 4449 [49198] "preview</w>"
  -1.6020185947418213 * 38467 [48474] "lindo</w>"
  -1.4871629476547241 * 1911 [49195] "kids</w>"
  -1.4275604486465454 * 5100 [49177] "potential</w>"
  1.3790473937988281 * 8222 [836] "statue</w>"
  1.4073306322097778 * 42901 [209] "calcutta</w>"
  1.4927211999893188 * 15035 [1710] "phenomenal</w>"
  1.5150572061538696 * 15697 [70] "sigma</w>"
  1.6294268369674683 * 19844 [754] "chero"
  1.6761832237243652 * 5834 [196] "heh</w>
  1.768609881401062 * 36700 [563] "âľĵ</w>"
  1.814394235610962 * 35575 [221] "kier</w>"
  1.8617881536483765 * 48095 [7] "bourdain</w>"
  1.937498688697815 * 38834 [1] "anjali</w>"
  2.204029083251953 * 8693 [188] "ori</w>"
  2.231095552444458 * 9326 [6] "pomegran"

[new-buddha-statue-1.bin.txt](https://github.com/tkalayci71/embedding-inspector/files/10356624/new-buddha-statue-1.bin.txt)

If you can't be bothered to try, don't worry, I'll try thus result (or more accurately, future, better-converged results) a week or two from now when I get my replacement CPU and can run SD again. At this point, I don't even know what the Buddha statue binary is supposed to look like, or what checkpoint it's supposed to be used with, or anything.  I did, however, try just punching the words (without any weight) into a free SD site, and I think it's en route to the right track, at least. Three runs, no cherry-picking:

![image](https://user-images.githubusercontent.com/116288799/210909877-eef2dc24-1575-41f7-9bf0-24feca4787e2.png)
![image](https://user-images.githubusercontent.com/116288799/210910111-97bc5205-2715-47a0-acb3-2a6944fd9820.png)
![image](https://user-images.githubusercontent.com/116288799/210910208-0dc7d732-14c3-46f7-9ee6-4c585363eaf2.png)

Funny how it's clearly picking up on the concept of Buddha, without the word "Buddha" in it.  Can't wait to see what it looks like with proper concept-mixing and further optimization  :)

**Comment by tkalayci71** [comment]
Good for you. I double checked with my embedding-mixer, some names are incorrect (#9326 is "displa"), but the mixed embedding is correct and it does produce similar images. Still, I don't feel excited, because the words are kind of random and it doesn't seem very meaningful or useful. Except, since output is independent of vector size, maybe 768 <->1024 converter can be made. So I'm still out, but happy for you.

**Comment by enn-nafnlaus** [comment]
I translated those by hand (that was generated before I incorporated your above code last night). I didn't know that you wanted specific tokens to be included in the output.  I can certainly let the user provide hints or mandated tokens and let it solve for what else might best fill in the gaps, if that would interest you.  That would probably help improve both overall convergence and convergence speed as well.
